src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1567,16 +1572,57 @@ internal new NamedTypeSymbol GetSpecialType(SpecialType specialType)
         }
 
         /// <summary>
-        /// Cache of T to Nullable&lt;T&gt;.
+        /// Given a provided <paramref name="typeArgument"/>, gives back <see cref="Nullable{T}"/> constructed with that
+        /// argument.  This function is only intended to be used for very common instantiations produced heavily during
+        /// binding.  Specifically, the nullable versions of enums, and the nullable versions of core built-ins.  So
+        /// many of these are created that it's worthwhile to cache, keeping overall garbage low, while not ballooning
+        /// the size of the cache itself.
         /// </summary>
-        private ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol> _typeToNullableVersion = ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol>.Empty;
-
         internal NamedTypeSymbol GetOrCreateNullableType(TypeSymbol typeArgument)
-            => RoslynImmutableInterlocked.GetOrAdd(
-                ref _typeToNullableVersion,
-                typeArgument,
-                static (typeArgument, @this) => @this.GetSpecialType(SpecialType.System_Nullable_T).Construct(typeArgument),
-                this);
+        {
+#if DEBUG
+            Debug.Assert(isSupportedType(typeArgument));
```

Comment:
Nit: I think it would be worth including the type that failed in the assertion for debugging purposes.